mypy/build.py

```diff
@@ -933,13 +937,26 @@ def _find_module(self, id: str, lib_path: Tuple[str, ...],
             typed_file = os.path.join(pkg_dir, components[0], 'py.typed')
             stub_dir = os.path.join(pkg_dir, stub_name)
             if fscache.isdir(stub_dir):
+                stub_typed_file = os.path.join(stub_dir, 'py.typed')
                 stub_components = [stub_name] + components[1:]
                 path = os.path.join(pkg_dir, *stub_components[:-1])
                 if fscache.isdir(path):
-                    third_party_stubs_dirs.append(path)
+                    if fscache.isfile(stub_typed_file):
+                        # Stub packages can have a py.typed file, which must include
+                        # 'partial\n' to make the package partial
+                        # Partial here means that mypy should look at the runtime
+                        # package if installed.
+                        if fscache.read(stub_typed_file).decode().strip() == 'partial':
+                            runtime_path = os.path.join(pkg_dir, dir_chain)
```

Comment:
PEP 561 is not 100% clear about this, but I think mypy should also use typeshed if a module is not found in a partial stub package. _Unless I am missing something_, the current implementation only captures one use case when a stub package is used to override an inline/runtime package. But there is another use case allowed by PEP 561 -- override typeshed with a stub package, what to do if the latter is declared partial and a module is not found? I think mypy should look in typeshed.

I mean you have this list 1...5 in the PEP, I think a simple rule is that if a module is not found in (3) but the package is partial we should just continue in normal order to (4) inline/runtime packages, _and then_ (5) typeshed.

Reply:
I agree the PEP text should be amended to be this way, fortunately for mypy, we almost already do what you describe. I *think* if I just re-order the elements of `candidate_base_dirs` to be in the right order as described by the PEP (which it sadly currently is not), it will "just work".

Reply:
PR is https://github.com/python/mypy/pull/5256

Reply:
I will merge this one now, and will review #5256 later (it looks independent).